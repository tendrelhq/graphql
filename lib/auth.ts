import { decodeGlobalId } from "@/schema/system";
import type { Context } from "@/schema/types";
import {
  ClerkExpressRequireAuth,
  type StrictAuthProp,
} from "@clerk/clerk-sdk-node";
import type e from "express";
import type { RequestHandler } from "express";
import { GraphQLError } from "graphql";
import * as jose from "jose";
import { type TxSql, sql } from "./datasources/postgres";
import { assert } from "./util";

declare global {
  namespace Express {
    interface Request extends StrictAuthProp {}
  }
}

export type Auth = StrictAuthProp["auth"];

export function clerk() {
  // In development it is nice to be able to use graphiql to help write your
  // queries. The default `devenv` will spin up a ruru[^1] graphiql endpoint
  // that you can use from your browser: http://localhost:1337.
  //
  // [^1]: https://github.com/graphile/crystal/tree/main/grafast/ruru
  //
  // tl;dr if you set the 'x-tendrel-user' header to a Clerk user id, we will
  // use that as the authenticated identity
  return (req: e.Request, res: e.Response, next: e.NextFunction) => {
    if (process.env.NODE_ENV === "development") {
      const userId = req.headers["x-tendrel-user"];
      console.log(`Backdoor hack engaged? ${userId}`);
      if (userId) {
        req.auth = {
          userId: userId as string,
        } as Auth;

        return next();
      }
    }

    ClerkExpressRequireAuth()(req, res, err => {
      if (err instanceof Error) {
        // @ts-ignore
        err.statusCode = 401;
      }

      next(err);
    });
  };
}

/**
 * Verify the provided *external* security token, generate a short-lived,
 * finely-grained subject token, and then call our /token api to perform the
 * actual token exchange. The client should not be aware that we are performing
 * this multi-step process.
 */
export const login: RequestHandler = async (req, res) => {
  // TODO: Move this elsewhere.
  const ALG = process.env.JWT_ALG ?? "HS256";
  const SECRET = new TextEncoder().encode(process.env.JWT_SECRET);
  // N.B. when JWT_ALG = HS256, JWT_SECRET must be at least 32 bytes!
  if (ALG === "HS256") assert(SECRET.length >= 32, "invalid jwt secret");

  const { userId } = req.auth;

  // TODO: ensure the user exists in the system, effectively making this API a
  // dual login/signup API. For now we will assume that the user has already
  // "signed up" for a Tendrel account, i.e. via the canonical Clerk/console flow.

  const temp = await new jose.SignJWT({ role: "anonymous" })
    .setProtectedHeader({ alg: ALG, typ: "JWT" })
    .setIssuer(`urn:tendrel:${process.env.STAGE}`)
    .setIssuedAt() // now
    .setNotBefore("-5 minutes")
    .setExpirationTime("5 minutes")
    .setSubject(userId)
    .sign(SECRET);

  try {
    // Same URL in both development and production. In development this will hit
    // the nginx proxy running on localhost. In production it will hit the same
    // proxy but running in a separate (ECS) container.
    const token = await fetch("http://localhost/api/v1/rpc/token", {
      method: "POST",
      body: JSON.stringify({
        grant_type: "urn:ietf:params:oauth:grant-type:token-exchange",
        subject_token: temp,
        subject_token_type: "urn:ietf:params:oauth:token-type:jwt",
      }),
      headers: {
        Authorization: `Bearer ${temp}`,
      },
    });

    const json = await token.json();

    res
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.1
      .setHeader("Cache-Control", "no-store")
      .setHeader("Pragma", "no-cache")
      .status(token.status);

    if (!token.ok) {
      // https://www.rfc-editor.org/rfc/rfc6749#section-5.2
      res.json({
        error: json.code,
        error_description: json.message,
      });
    } else {
      res.json(json);
    }
  } catch (e) {
    console.error(e);
    res.status(500).send("Internal Server Error");
  }
};

export async function protect(
  { orgId, userId }: { orgId: string; userId: string },
  allowed: ("Worker" | "Supervisor" | "Admin")[],
) {
  const rows = await sql`
    select 1
    from public.workerinstance as w
    inner join public.systag as s on w.workerinstanceuserroleid = s.systagid
    where
        w.workerinstancecustomerid = (
            select customerid
            from public.customer
            where customeruuid = ${decodeGlobalId(orgId).id}
        )
        and w.workerinstanceworkerid = (
            select workerid
            from public.worker
            where workeridentityid = ${userId}
        )
        and s.systagtype in (${sql(allowed)});
  `;

  if (!rows.length) {
    throw new GraphQLError("Not authorized", {
      extensions: {
        code: "UNAUTHORIZED",
        hint: "You do not have the necessary permissions to perform this action",
      },
    });
  }
}

export function setCurrentIdentity(sql: TxSql, ctx: Context) {
  return sql`select * from auth.set_actor(${ctx.auth.userId}, ${ctx.req.i18n.language})`;
}
