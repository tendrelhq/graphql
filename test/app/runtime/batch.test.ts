import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { setCurrentIdentity } from "@/auth";
import { sql } from "@/datasources/postgres";
import { schema } from "@/schema/final";
import type { Location } from "@/schema/platform/archetype/location";
import type { Task } from "@/schema/system/component/task";
import {
  type Customer,
  assertTaskIsNamed,
  assertTaskParentIs,
  createEmptyCustomer,
  createTestContext,
  execute,
  getFieldByName,
} from "@/test/prelude";
import { assert, assertNonNull } from "@/util";
import {
  AssignBatchMutationDocument,
  CreateBatchMutationDocument,
  TestBatchEntrypointDocument,
} from "./batch.test.generated";
import { getLatestFsm, mostRecentInstance } from "./prelude";
import {
  TestRuntimeDetailDocument,
  TestRuntimeEntrypointDocument,
  TestRuntimeTransitionMutationDocument,
} from "./runtime.test.generated";

const ctx = await createTestContext();

describe("runtime + batch tracking", () => {
  // See beforeAll for initialization of these variables.
  let CUSTOMER: Customer;
  let BATCH_TEMPLATE: Task;
  let FACTORY: Location;
  let MIXING_LINE: Location;
  let FILL_LINE: Location;
  let RUN_TEMPLATE: Task;

  test("no batches at first", async () => {
    const result = await execute(schema, TestBatchEntrypointDocument, {
      parent: CUSTOMER.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data?.trackables?.edges?.length).toBe(0);
  });

  let batchId = 0;

  test("create a new batch", async () => {
    // Note that this only creates an *open* Batch instance. In particular, this
    // would not show up on the current Runtime home screen.
    const customer = await getFieldByName(BATCH_TEMPLATE, "Customer");
    const productName = await getFieldByName(BATCH_TEMPLATE, "Product Name");
    const sku = await getFieldByName(BATCH_TEMPLATE, "SKU");
    const result = await execute(schema, CreateBatchMutationDocument, {
      batchTemplateId: BATCH_TEMPLATE.id,
      batchId: (batchId++).toString(),
      fields: [
        {
          field: customer.id,
          valueType: customer.valueType,
          value: { string: "Ross's Salsa" },
        },
        {
          field: productName.id,
          valueType: productName.valueType,
          value: { string: "Mild Green Salsa" },
        },
        {
          field: sku.id,
          valueType: sku.valueType,
          value: { string: "SLS-GRN-ML" },
        },
      ],
      location: FACTORY.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("still no batches via the canonical entrypoint query", async () => {
    const result = await execute(schema, TestRuntimeEntrypointDocument, {
      parent: CUSTOMER.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("but it shows up via the batch entrypoint query", async () => {
    const result = await execute(schema, TestBatchEntrypointDocument, {
      parent: CUSTOMER.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("assign the batch", async () => {
    const node = await RUN_TEMPLATE.instantiate(
      {
        location: MIXING_LINE.id,
      },
      ctx,
    );
    const base = await mostRecentInstance(BATCH_TEMPLATE);
    expect(node.id).not.toBe(base.id);

    const result = await execute(schema, AssignBatchMutationDocument, {
      node: node.id,
      base: base.id,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchObject({
      rebase: {
        name: {
          value: "Run",
        },
        parent: {
          name: {
            value: "Mixing Line",
          },
        },
        root: {
          name: {
            value: "Batch",
          },
        },
        state: {
          __typename: "Open",
        },
      },
    });
  });

  test("start the batch", async () => {
    const t = await mostRecentInstance(BATCH_TEMPLATE);
    const h = await t.hash();
    const result = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: false,
        includeTransitionIds: true,
        opts: {
          fsm: {
            id: t.id,
            hash: h,
          },
          task: {
            id: t.id,
            hash: h,
          },
        },
      },
    );
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchObject({
      advance: {
        root: {
          fsm: {
            active: {
              name: {
                value: "Run",
              },
              parent: {
                name: {
                  value: "Mixing Line",
                },
              },
              state: {
                __typename: "Open",
              },
            },
          },
        },
      },
    });
  });

  test("start run @ mixing line", async () => {
    // We should already have an *open* Run instance at the Mixing Line because
    // we rebased onto the Batch in an earlier test!
    const { fsm, root } = await getLatestFsm(BATCH_TEMPLATE);
    const active = assertNonNull(fsm.active, "should be active");
    assertTaskIsNamed(active, "Run", ctx);
    assertTaskParentIs(active, MIXING_LINE);

    const startRun = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: true,
        opts: {
          fsm: {
            id: root.id,
            hash: await root.hash(),
          },
          task: {
            id: active.id,
            hash: await active.hash(),
          },
        },
      },
    );
    expect(startRun.errors).toBeFalsy();
    expect(startRun.data?.advance?.root?.fsm?.active).toMatchSnapshot();
  });

  test("end run @ mixing line", async () => {
    const { fsm, root } = await getLatestFsm(BATCH_TEMPLATE);
    const active = assertNonNull(fsm.active, "should be active");
    assertTaskIsNamed(active, "Run", ctx);

    const finishRun = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: true,
        opts: {
          fsm: {
            id: root.id,
            hash: await root.hash(),
          },
          task: {
            id: active.id,
            hash: await active.hash(),
          },
        },
      },
    );
    expect(finishRun.errors).toBeFalsy();
    expect(finishRun.data?.advance?.root?.fsm?.active).toMatchSnapshot();
    expect(finishRun.data?.advance?.root?.chain).toMatchSnapshot();
  });

  test("start run @ fill line", async () => {
    // The engine will have created a Run instance at the Fill Line, as per the
    // transitions we configure in the `beforeAll` phase.
    const { fsm, root } = await getLatestFsm(BATCH_TEMPLATE);
    const active = assertNonNull(fsm.active, "should be active");
    assertTaskIsNamed(active, "Run", ctx);
    assertTaskParentIs(active, FILL_LINE);

    const startRun = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: true,
        opts: {
          fsm: {
            id: root.id,
            hash: await root.hash(),
          },
          task: {
            id: active.id,
            hash: await active.hash(),
          },
        },
      },
    );
    expect(startRun.errors).toBeFalsy();
    expect(startRun.data?.advance?.root?.fsm?.active).toMatchSnapshot();
  });

  test("end run @ fill line", async () => {
    const { fsm, root } = await getLatestFsm(BATCH_TEMPLATE);
    const active = assertNonNull(fsm.active, "should be active");
    assertTaskIsNamed(active, "Run", ctx);

    const finishRun = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: true,
        opts: {
          fsm: {
            id: root.id,
            hash: await root.hash(),
          },
          task: {
            id: active.id,
            hash: await active.hash(),
          },
        },
      },
    );
    expect(finishRun.errors).toBeFalsy();
    expect(finishRun.data?.advance?.root?.fsm?.active).toMatchSnapshot();
    expect(finishRun.data?.advance?.root?.chain).toMatchSnapshot();
  });

  test("close out the batch", async () => {
    const { root } = await getLatestFsm(BATCH_TEMPLATE);
    const rootHash = await root.hash();
    const result = await execute(
      schema,
      TestRuntimeTransitionMutationDocument,
      {
        includeChain: false,
        opts: {
          fsm: {
            id: root.id,
            hash: rootHash,
          },
          task: {
            id: root.id,
            hash: rootHash,
          },
        },
      },
    );
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchSnapshot();
  });

  test("closed batch view", async () => {
    const i = await mostRecentInstance(BATCH_TEMPLATE);
    const result = await execute(schema, TestRuntimeDetailDocument, {
      node: i.id,
      overTypes: ["Batch", "Runtime"],
      includeChainParents: true,
    });
    expect(result.errors).toBeFalsy();
    expect(result.data).toMatchObject({
      node: {
        chain: {
          edges: [
            {
              node: {
                name: {
                  value: "Batch", // FIXME
                },
                parent: {
                  name: {
                    value: "Frozen Tendy Factory",
                  },
                },
                state: {
                  __typename: "Closed",
                },
              },
            },
            {
              node: {
                name: {
                  value: "Run",
                },
                parent: {
                  name: {
                    value: "Mixing Line",
                  },
                },
                state: {
                  __typename: "Closed",
                },
              },
            },
            {
              node: {
                name: {
                  value: "Run",
                },
                parent: {
                  name: {
                    value: "Fill Line",
                  },
                },
                state: {
                  __typename: "Closed",
                },
              },
            },
            // TODO: Cascade when operating on the root?
            // Or, ideally, any ancestor.
            {
              node: {
                name: {
                  value: "Run",
                },
                parent: {
                  name: {
                    value: "Assembly Line",
                  },
                },
                state: {
                  __typename: "Open",
                },
              },
            },
          ],
          totalCount: 4,
        },
        chainAgg: [
          {
            group: "Batch",
            value: expect.stringMatching(/\d+.\d+/),
          },
          {
            group: "Runtime",
            value: expect.stringMatching(/\d+.\d+/),
          },
        ],
        parent: {
          name: {
            value: "Frozen Tendy Factory",
          },
        },
      },
    });
  });

  beforeAll(async () => {
    // Setup:
    await sql.begin(async sql => {
      await setCurrentIdentity(sql, ctx);

      CUSTOMER = await createEmptyCustomer(
        { name: "Frozen Tendy Factory" },
        ctx,
        sql,
      );

      await CUSTOMER.addWorker(
        { identityId: assertNonNull(process.env.X_TENDREL_USER) },
        ctx,
        sql,
      );

      // Setup "Runtime" and configure "Batch tracking".
      // This involves:
      // 1. Locations: site + Mixing, Fill, Assembly, Cartoning and Packaging Lines
      // 2. Templates:
      //   - Batch: instantiated at the site level
      //   - Run, Down, Idle: instantiated at the "Lines" level
      // 3. Rules:
      //   - *No* in-progress/respawn rule. Batches (instances) must be created
      //     manually.
      //   - When a Run instance is Closed, an Open Run instance should be
      //     [eagerly] created at the "next" line. Same originator (i.e. Batch).

      // Locations.
      FACTORY = await CUSTOMER.addLocation(
        { name: "Frozen Tendy Factory", type: "Frozen Tendy Factory" },
        ctx,
        sql,
      );
      MIXING_LINE = await FACTORY.insertChild(
        { name: "Mixing Line", order: 0, type: "Runtime Location" },
        ctx,
        sql,
      );
      FILL_LINE = await FACTORY.insertChild(
        { name: "Fill Line", order: 1, type: "Runtime Location" },
        ctx,
        sql,
      );
      const assemblyLine = await FACTORY.insertChild(
        { name: "Assembly Line", order: 2, type: "Runtime Location" },
        ctx,
        sql,
      );
      const cartoningLine = await FACTORY.insertChild(
        { name: "Cartoning Line", order: 3, type: "Runtime Location" },
        ctx,
        sql,
      );
      const packagingLine = await FACTORY.insertChild(
        { name: "Packaging Line", order: 4, type: "Runtime Location" },
        ctx,
        sql,
      );

      // Templates.
      BATCH_TEMPLATE = await FACTORY.createTemplate(
        {
          name: "Batch",
          fields: [
            { name: "Customer", type: "string" },
            { name: "Product Name", type: "string" },
            { name: "SKU", type: "string" },
          ],
          types: ["Batch"],
        },
        ctx,
        sql,
      );
      RUN_TEMPLATE = await FACTORY.createTemplate(
        {
          name: "Run",
          fields: [
            // Note that we still have the primary/order requirement for
            // Overrides (primary + 0 => start, 1 => end).
            // This will be removed in the future.
            {
              name: "Override Start Time",
              type: "timestamp",
              isPrimary: true,
              order: 0,
            },
            {
              name: "Override End Time",
              type: "timestamp",
              isPrimary: true,
              order: 1,
            },
            { name: "Run Output", type: "number", order: 2 },
            { name: "Reject Count", type: "number", order: 3 },
            { name: "Comments", type: "string", order: 99 },
          ],
          types: ["Trackable", "Runtime"],
        },
        ctx,
        sql,
      );
      const downTemplate = await FACTORY.createTemplate(
        {
          name: "Downtime",
          fields: [
            {
              name: "Override Start Time",
              type: "timestamp",
              isPrimary: true,
              order: 0,
            },
            {
              name: "Override End Time",
              type: "timestamp",
              isPrimary: true,
              order: 1,
            },
            { name: "Description", type: "string", order: 99 },
          ],
          types: ["Downtime"],
        },
        ctx,
        sql,
      );
      const idleTemplate = await FACTORY.createTemplate(
        {
          name: "Idle Time",
          fields: [
            {
              name: "Override Start Time",
              type: "timestamp",
              isPrimary: true,
              order: 0,
            },
            {
              name: "Override End Time",
              type: "timestamp",
              isPrimary: true,
              order: 1,
            },
            { name: "Description", type: "string", order: 99 },
          ],
          types: ["Idle Time"],
        },
        ctx,
        sql,
      );

      // Constraints.
      await BATCH_TEMPLATE.ensureInstantiableAt(
        { locations: [FACTORY.id] },
        ctx,
        sql,
      );
      await BATCH_TEMPLATE.ensureInstantiableAt(
        { locations: [FACTORY.id] },
        ctx,
        sql,
      );
      for (const t of [RUN_TEMPLATE, downTemplate, idleTemplate]) {
        await t.ensureInstantiableAt(
          {
            locations: [
              MIXING_LINE.id,
              FILL_LINE.id,
              assemblyLine.id,
              cartoningLine.id,
              packagingLine.id,
            ],
          },
          ctx,
          sql,
        );
      }

      // Rules.
      // When the Batch goes in-progress, prompt the user to choose where it
      // should go next. For the purposes of this test, we will allow for the
      // Batch to start at any Line.
      for (const l of [
        MIXING_LINE,
        FILL_LINE,
        assemblyLine,
        cartoningLine,
        packagingLine,
      ]) {
        await BATCH_TEMPLATE.createTransition(
          {
            whenStatusChangesTo: "InProgress",
            instantiate: {
              template: RUN_TEMPLATE.id,
              atLocation: l.id,
              withType: "On Demand", // Choice.
            },
          },
          ctx,
          sql,
        );
      }

      // Mixing -> Fill
      await RUN_TEMPLATE.createTransition(
        {
          atLocation: MIXING_LINE.id,
          whenStatusChangesTo: "Closed",
          instantiate: {
            template: RUN_TEMPLATE.id,
            atLocation: FILL_LINE.id,
            withType: "Task", // Eager. This is the default.
          },
        },
        ctx,
        sql,
      );

      // Fill -> Assembly
      await RUN_TEMPLATE.createTransition(
        {
          atLocation: FILL_LINE.id,
          whenStatusChangesTo: "Closed",
          instantiate: {
            template: RUN_TEMPLATE.id,
            atLocation: assemblyLine.id,
          },
        },
        ctx,
        sql,
      );

      // Assembly -> Cartoning
      await RUN_TEMPLATE.createTransition(
        {
          atLocation: assemblyLine.id,
          whenStatusChangesTo: "Closed",
          instantiate: {
            template: RUN_TEMPLATE.id,
            atLocation: cartoningLine.id,
          },
        },
        ctx,
        sql,
      );

      // Cartoning -> Packaging
      await RUN_TEMPLATE.createTransition(
        {
          atLocation: cartoningLine.id,
          whenStatusChangesTo: "Closed",
          instantiate: {
            template: RUN_TEMPLATE.id,
            atLocation: packagingLine.id,
          },
        },
        ctx,
        sql,
      );

      // FIXME: use Keller's API for customer create through the entity model.
      // await sql`call entity.import_entity(null)`;
    });
  });

  afterAll(async () => {
    // Cleanup:
    // await cleanup(CUSTOMER);
  });
});
