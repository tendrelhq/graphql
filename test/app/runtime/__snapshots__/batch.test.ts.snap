// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`runtime + batch tracking create a new batch 1`] = `
{
  "createInstance": {
    "__typename": "CreateInstancePayload",
    "edge": {
      "__typename": "EntityInstanceEdge",
      "node": {
        "__typename": "EntityInstance",
        "asTask": {
          "__typename": "Task",
          "name": {
            "__typename": "DisplayName",
            "value": "Batch",
          },
          "parent": {
            "__typename": "Location",
            "name": {
              "__typename": "Name",
              "value": "Frozen Tendy Factory",
            },
          },
          "state": {
            "__typename": "Open",
          },
        },
      },
    },
  },
}
`;

exports[`runtime + batch tracking still no batches via the canonical entrypoint query 1`] = `
{
  "trackables": {
    "__typename": "TrackableConnection",
    "edges": [
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Location",
          "name": {
            "__typename": "Name",
            "value": "Mixing Line",
          },
          "timeZone": "utc",
          "tracking": {
            "edges": [],
          },
        },
      },
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Location",
          "name": {
            "__typename": "Name",
            "value": "Fill Line",
          },
          "timeZone": "utc",
          "tracking": {
            "edges": [],
          },
        },
      },
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Location",
          "name": {
            "__typename": "Name",
            "value": "Assembly Line",
          },
          "timeZone": "utc",
          "tracking": {
            "edges": [],
          },
        },
      },
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Location",
          "name": {
            "__typename": "Name",
            "value": "Cartoning Line",
          },
          "timeZone": "utc",
          "tracking": {
            "edges": [],
          },
        },
      },
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Location",
          "name": {
            "__typename": "Name",
            "value": "Packaging Line",
          },
          "timeZone": "utc",
          "tracking": {
            "edges": [],
          },
        },
      },
    ],
    "totalCount": 5,
  },
}
`;

exports[`runtime + batch tracking but it shows up via the batch entrypoint query 1`] = `
{
  "trackables": {
    "__typename": "TrackableConnection",
    "edges": [
      {
        "__typename": "TrackableEdge",
        "node": {
          "__typename": "Task",
          "fields": {
            "__typename": "FieldConnection",
            "edges": [
              {
                "__typename": "FieldEdge",
                "node": {
                  "__typename": "Field",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Customer",
                  },
                },
              },
              {
                "__typename": "FieldEdge",
                "node": {
                  "__typename": "Field",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "Product Name",
                  },
                },
              },
              {
                "__typename": "FieldEdge",
                "node": {
                  "__typename": "Field",
                  "name": {
                    "__typename": "DisplayName",
                    "value": "SKU",
                  },
                },
              },
            ],
          },
          "fsm": {
            "__typename": "TaskStateMachine",
            "active": {
              "__typename": "Task",
              "parent": {
                "__typename": "Location",
                "name": {
                  "__typename": "Name",
                  "value": "Frozen Tendy Factory",
                },
              },
            },
            "transitions": {
              "__typename": "TaskTransitions",
              "edges": [],
            },
          },
          "name": {
            "__typename": "DisplayName",
            "value": "Batch",
          },
        },
      },
    ],
  },
}
`;

exports[`runtime + batch tracking start run @ mixing line 1`] = `
{
  "assignees": {
    "edges": [
      {
        "node": {
          "assignedTo": {
            "displayName": "Jerry Garcia",
          },
        },
      },
    ],
  },
  "description": null,
  "fields": {
    "edges": [
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override Start Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override End Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Run Output",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Reject Count",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Comments",
          },
          "value": {
            "__typename": "StringValue",
            "string": null,
          },
          "valueType": "string",
        },
      },
    ],
  },
  "name": {
    "value": "Run",
  },
  "parent": {
    "name": {
      "value": "Mixing Line",
    },
  },
  "state": {
    "__typename": "InProgress",
    "inProgressBy": {
      "displayName": "Jerry Garcia",
    },
  },
}
`;

exports[`runtime + batch tracking end run @ mixing line 1`] = `
{
  "assignees": {
    "edges": [
      {
        "node": {
          "assignedTo": null,
        },
      },
    ],
  },
  "description": null,
  "fields": {
    "edges": [
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override Start Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override End Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Run Output",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Reject Count",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Comments",
          },
          "value": {
            "__typename": "StringValue",
            "string": null,
          },
          "valueType": "string",
        },
      },
    ],
  },
  "name": {
    "value": "Run",
  },
  "parent": {
    "name": {
      "value": "Fill Line",
    },
  },
  "state": {
    "__typename": "Open",
  },
}
`;

exports[`runtime + batch tracking end run @ mixing line 2`] = `
{
  "__typename": "TaskConnection",
  "edges": [
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Batch",
        },
        "state": {
          "__typename": "InProgress",
        },
      },
    },
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Run",
        },
        "state": {
          "__typename": "Closed",
        },
      },
    },
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Run",
        },
        "state": {
          "__typename": "Open",
        },
      },
    },
  ],
}
`;

exports[`runtime + batch tracking start run @ fill line 1`] = `
{
  "assignees": {
    "edges": [
      {
        "node": {
          "assignedTo": {
            "displayName": "Jerry Garcia",
          },
        },
      },
    ],
  },
  "description": null,
  "fields": {
    "edges": [
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override Start Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override End Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Run Output",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Reject Count",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Comments",
          },
          "value": {
            "__typename": "StringValue",
            "string": null,
          },
          "valueType": "string",
        },
      },
    ],
  },
  "name": {
    "value": "Run",
  },
  "parent": {
    "name": {
      "value": "Fill Line",
    },
  },
  "state": {
    "__typename": "InProgress",
    "inProgressBy": {
      "displayName": "Jerry Garcia",
    },
  },
}
`;

exports[`runtime + batch tracking end run @ fill line 1`] = `
{
  "assignees": {
    "edges": [
      {
        "node": {
          "assignedTo": null,
        },
      },
    ],
  },
  "description": null,
  "fields": {
    "edges": [
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override Start Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Override End Time",
          },
          "value": {
            "__typename": "TimestampValue",
            "timestamp": null,
          },
          "valueType": "timestamp",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Run Output",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Reject Count",
          },
          "value": {
            "__typename": "NumberValue",
          },
          "valueType": "number",
        },
      },
      {
        "node": {
          "description": null,
          "name": {
            "value": "Comments",
          },
          "value": {
            "__typename": "StringValue",
            "string": null,
          },
          "valueType": "string",
        },
      },
    ],
  },
  "name": {
    "value": "Run",
  },
  "parent": {
    "name": {
      "value": "Assembly Line",
    },
  },
  "state": {
    "__typename": "Open",
  },
}
`;

exports[`runtime + batch tracking end run @ fill line 2`] = `
{
  "__typename": "TaskConnection",
  "edges": [
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Batch",
        },
        "state": {
          "__typename": "InProgress",
        },
      },
    },
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Run",
        },
        "state": {
          "__typename": "Closed",
        },
      },
    },
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Run",
        },
        "state": {
          "__typename": "Closed",
        },
      },
    },
    {
      "__typename": "TaskEdge",
      "node": {
        "__typename": "Task",
        "name": {
          "__typename": "DisplayName",
          "value": "Run",
        },
        "state": {
          "__typename": "Open",
        },
      },
    },
  ],
}
`;

exports[`runtime + batch tracking close out the batch 1`] = `
{
  "advance": {
    "__typename": "AdvanceTaskStateMachineResult",
    "diagnostics": null,
    "instantiations": [],
    "root": {
      "__typename": "Task",
      "fsm": {
        "__typename": "TaskStateMachine",
        "active": {
          "assignees": {
            "edges": [
              {
                "node": {
                  "assignedTo": null,
                },
              },
            ],
          },
          "description": null,
          "fields": {
            "edges": [
              {
                "node": {
                  "description": null,
                  "name": {
                    "value": "Override Start Time",
                  },
                  "value": {
                    "__typename": "TimestampValue",
                    "timestamp": null,
                  },
                  "valueType": "timestamp",
                },
              },
              {
                "node": {
                  "description": null,
                  "name": {
                    "value": "Override End Time",
                  },
                  "value": {
                    "__typename": "TimestampValue",
                    "timestamp": null,
                  },
                  "valueType": "timestamp",
                },
              },
              {
                "node": {
                  "description": null,
                  "name": {
                    "value": "Run Output",
                  },
                  "value": {
                    "__typename": "NumberValue",
                  },
                  "valueType": "number",
                },
              },
              {
                "node": {
                  "description": null,
                  "name": {
                    "value": "Reject Count",
                  },
                  "value": {
                    "__typename": "NumberValue",
                  },
                  "valueType": "number",
                },
              },
              {
                "node": {
                  "description": null,
                  "name": {
                    "value": "Comments",
                  },
                  "value": {
                    "__typename": "StringValue",
                    "string": null,
                  },
                  "valueType": "string",
                },
              },
            ],
          },
          "name": {
            "value": "Run",
          },
          "parent": {
            "name": {
              "value": "Assembly Line",
            },
          },
          "state": {
            "__typename": "Open",
          },
        },
        "transitions": {
          "edges": [],
        },
      },
      "state": {
        "__typename": "Closed",
      },
    },
  },
}
`;
